src/Collection.ts

```diff
@@ -3,19 +3,59 @@ import Model from './Model'
 import Request from './Request'
 import apiClient from './apiClient'
 import difference from 'lodash/difference'
-import isMatch from 'lodash/isMatch'
-import { observable, action, computed, IObservableArray } from 'mobx'
-
+import intersection from 'lodash/intersection'
+import entries from 'lodash/entries'
+import { observable, action, computed, IObservableArray, reaction } from 'mobx'
 import { CreateOptions, SetOptions, GetOptions, FindOptions, Id } from './types'
 
+type IndexTree<T> = Map<string, Index<T>>
+type Index<T> = Map<any, Array<T>>
+
 export default abstract class Collection<T extends Model> extends Base {
   @observable models: IObservableArray<T> = observable.array([])
+  indexes: Array<string> = []
 
   constructor (data: Array<{ [key: string]: any }> = []) {
     super()
     this.set(data)
   }
 
+  /**
+   * Define which is the primary key
+   * of the models in the collection.
+   *
+   * FIXME: This contains a hack to use the `primaryKey` as
+   * an instance method. Ideally it should be static but that
+   * would not be backward compatible and Typescript sucks at
+   * static polymorphism (https://github.com/microsoft/TypeScript/issues/5863).
+   */
+  get primaryKey (): string {
+    const ModelClass = this.model()
+    return (new ModelClass()).primaryKey
+  }
+
+  /**
+   * Returns a hash with all the indexes for that
+   * collection.
+   */
+  @computed({ keepAlive: true })
+  get index (): IndexTree<T> {
+    const indexes = this.indexes.concat([this.primaryKey])
+
+    return indexes.reduce((tree, attr) => {
+      const newIndex = this.models.reduce((index: Index<T>, model: T) => {
+        const value = model.has(attr)
+          ? model.get(attr)
+          : null
+        const oldModels = index.get(value) || []
```

Comment:
Which is the expected behaviour when the value is `null`? Maybe an early return de ignoring this item.

Reply:
ideally, I would early return. Unluckily, javascript does not have a way to do so other than exceptions. In ruby you have `next` and `return` which gives you the mechanism to do so.

Reply:
Add those `null` and `undefined` paths in tests

Reply:
done